src/task_impl/atomic_task.rs

```diff
@@ -86,90 +73,96 @@ impl AtomicTask {
     /// idea. Concurrent calls to `register` will attempt to register different
     /// tasks to be notified. One of the callers will win and have its task set,
     /// but there is no guarantee as to which caller will succeed.
-    pub fn register(&self) {
-        // Get a new task handle
-        let task = super::current();
-
-        match self.state.compare_and_swap(WAITING, LOCKED_WRITE, Acquire) {
+    pub fn register_task(&self, task: Task) {
+        match self.state.compare_and_swap(WAITING, REGISTERING, Acquire) {
             WAITING => {
                 unsafe {
-                    // Locked acquired, update the task cell
-                    *self.task.get() = Some(task);
-
-                    // Release the lock. If the state transitioned to
-                    // `LOCKED_NOTIFIED`, this means that an notify has been
-                    // signaled, so notify the task.
-                    if LOCKED_WRITE_NOTIFIED == self.state.swap(WAITING, Release) {
-                        (*self.task.get()).as_ref().unwrap().notify();
+                    // Locked acquired, update the waker cell
+                    *self.task.get() = Some(task.clone());
+
+                    // Release the lock. If the state transitioned to include
+                    // the `NOTIFYING` bit, this means that a notify has been
+                    // called concurrently, so we have to remove the task and
+                    // notify it.`
+                    //
+                    // Start by assuming that the state is `REGISTERING` as this
+                    // is what we jut set it to.
+                    let mut curr = REGISTERING;
+
+                    // If a task has to be notified, it will be set here.
+                    let mut notify = None;
+
+                    loop {
+                        let res = self.state.compare_exchange(
+                            curr, WAITING, AcqRel, Acquire);
```

Comment:
Should this be `Release, Acquire`? On success, we just need the previous writes to `self.task` to be written. On failure, we only need to be able to read others' writes to `self.task`.

Reply:
It has to be `AcqRel` as the success ordering must be stronger than the failure ordering (i.e. a super set).

Reply:
TIL! Thanks.